hid: tmff2: add Thrustmaster T500RS wheel base driver

cazzoo opened the pull request:
This driver has been built from scratch based the previous dirty driver (see https://github.com/Kimplul/hid-tmff2/pull/175) and on captures from ffbsdl tool ran in windows for all possible effects (through SDL2 library).

cazzoo committed:
hid: tmff2: add Thrustmaster T500RS wheel base driver

Add support for the Thrustmaster T500RS wheel base in the hid-tmff2 driver.

This driver has been built from scratch based the previous dirty driver (see https://github.com/Kimplul/hid-tmff2/pull/175) on captures from ffbsdl tool ran in windows for all possible effects (through SDL2 library).

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,2073 @@
+                 T500RS_MAX_HW_EFFECTS); /* occupied slots bitmap */
+
+  /*
+   * Thread safety - Phase 5 Security and Robustness.
```
What are these phases?

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,2073 @@
+    pr_warn_once("t500rs: Envelope requested but not supported for this effect type\n");
+  }
+
+  p->reserved = 0x00;
```
Since the packet is zero-initialized anyway, this is redundant.

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,2073 @@
+      }
+      struct t500rs_pkt_r05_condition *p =
+          (struct t500rs_pkt_r05_condition *)buf;
+      t500rs_build_r05_condition(p, (u8)(param_sub & 0xff), saturation);
```
Nitpick: `& 0xff` is not necessary when casting to `u8`.

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,2073 @@
+  duration_ms = effect->replay.length ? effect->replay.length : 0xffff;
+  delay_ms = effect->replay.delay;
+  period_ms = effect->u.periodic.period;
+  if (period_ms == 0)
```
I've brought this up a few times, but a default period length is not correct, as far as I'm aware. Just remove it or report an error or something instead.

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,2073 @@
+  init_buf[0] = 0x05;
+  init_buf[1] = 0x1c;
+  init_buf[2] = 0x00;
+  init_buf[3] = 0x00;  /* Deadband = 0 */
```
Why does the wheel have slots for deadband and center (assuming this comment is accurate) if they must be 0?

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
Redundant.

cazzoo replied:
Indeed, duplicate and non-necessary code

Kimplul commented on src/tmt500rs/t500rs_protocol.h:
```diff
@@ -0,0 +1,222 @@
+/* Sequence templates are now static in the implementation file */
+
+/* Supported parameters and effects */
+extern const unsigned long t500rs_params;
```
These should just be `static` in `hid-tmt500rs.c`.

Kimplul commented on src/hid-tmff2.c:
```diff
@@ -242,8 +242,31 @@ static ssize_t gain_store(struct device *dev,
+	*
+	* References:
+	* - docs/FFBEFFECTS.md: section "FF_GAIN" shows a dedicated device gain path.
+	* - docs/FFB_T500RS.md: Report glossary mentions 0x43 (gain), i.e., device-side
```
`docs/FFB_T500RS.md` doesn't exist.

cazzoo replied:
From old doc file, sorted!

Kimplul commented on src/hid-tmff2.c:
```diff
@@ -242,8 +242,31 @@ static ssize_t gain_store(struct device *dev,
+	* - docs/FFB_T500RS.md: Report glossary mentions 0x43 (gain), i.e., device-side
+	*   gain separate from per-effect magnitudes; drivers should expose both levels.
+	*/
+	if (tmff2->set_gain) {
```
Nitpick: this might be easier to read if you do `if (!tmff2->set_gain) return count;` and then do everything else.

cazzoo replied:
Good point, updated

Kimplul commented on src/hid-tmff2.c:
```diff
@@ -242,8 +242,31 @@ static ssize_t gain_store(struct device *dev,
+		unsigned long flags;
+		spin_lock_irqsave(&tmff2->lock, flags);
+		tmff2->pending_gain_value = GAIN_MAX;
+		tmff2->gain_pending = 1;
```
I mentioned this in the previous PR but I'd really prefer it if you created a new `FF_EFFECT_QUEUE_XXX` flag instead of a separate `gain_pending` variable. Same goes for  autocentering.

Kimplul commented on src/hid-tmff2.c:
```diff
@@ -315,15 +376,26 @@ static void tmff2_work_handler(struct work_struct *w)
 				__clear_bit(FF_EFFECT_QUEUE_UPDATE, &state->flags);
-
+				/* Request a STOP in process context */
+				__set_bit(FF_EFFECT_QUEUE_STOP, &actions);
```
Why is this necessary if the wheel has a `duration` field?

Kimplul commented on docs/T500RS_USB_Protocol_Analysis.md:
```diff
@@ -0,0 +1,663 @@
+
+**Sequential Effect Notes:**
+- Each effect gets full upload sequence with unique effect ID
+- Windows driver uses non-sequential IDs: 0x00, 0x21, 0x22, 0x40, 0x41
```
What are these IDs? effect IDs only go up to 15, don't they?

Kimplul reviewed (changes requested):
Thanks for continuing on this project, I was getting worried that we scared you off :)

I did a quick overview. The documentation in particular looks a lot better, having clear examples of what each packet does and how they build up to a full effect upload helped greatly. I haven't done any cross-referencing between the docs and implementation, as I still found quite a lot of smaller issues I'd like taken care of first. I'll do a more thorough review after that.

Not entirely sure what you mean by 'from scratch', for instance the comment explaining the `set_gain` rationale is identical between this and the previous PR.

cazzoo commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,2073 @@
+    return -EINVAL;
+  }
+
+  t500rs = data;
```
Good catch

cazzoo commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,2073 @@
+
+  /* Allocate send buffer with bounds checking */
+  t500rs->buffer_length = T500RS_BUFFER_LENGTH;
+  if (t500rs->buffer_length == 0 || t500rs->buffer_length > 4096) {
```
Removed

cazzoo commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,2073 @@
+  }
+
+  /* Validate gain range */
+  if (gain > T500RS_GAIN_MAX) {
```
Alright, removed

cazzoo commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,2073 @@
+  int hw_id;
+
+  /* Input validation */
+  if (!data) {
```
I reviewed most of the input validation and kept only the ones I considered most important

cazzoo commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,2073 @@
+    p->attack_level = t500rs_scale_envelope_level(env->attack_level);
+    p->fade_len = cpu_to_le16(env->fade_length);
+    p->fade_level = t500rs_scale_envelope_level(env->fade_level);
+  } else if (env && !allow_nonzero) {
```
Simplified

cazzoo commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,2073 @@
+                              effect->u.ramp.start_level,
+                              effect->u.ramp.end_level, effect->replay.length);
+      } else {
+        u8 mag = t500rs_scale_periodic_magnitude(effect->u.periodic.magnitude);
```
Good catch, I re-implemented logic done in T300RS and it appears to be fine

cazzoo committed:
Major pass of improvements based on PR feedback

- Simplified effect slot system
- Added support of direction
- Tidy-up code here and there
- Reduced complexity overhead
- Code standardization (in regards of the base repo)
- Documentation updates

cazzoo commented on src/tmt500rs/t500rs_protocol.h:
```diff
@@ -0,0 +1,222 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
```
Renamed! that's indeed more clean

cazzoo commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,2073 @@
+   * centering force which masks/overpowers other forces. To avoid this, ignore
+   * requests that try to set maximum autocenter (100%). Disabling (0) is still
+   * honored; lower values are allowed. */
+  if (autocenter_percent >= 100) {
```
Just double checked and reproduced the issue with RBR (RSF version). I don't know yet if that is a Wine bug or else, but this is really odd

Kimplul replied:
I think this should be removed, as I don't think it's the driver's responsibility to fix broken games. If the autocentering is too strong, it can be turned down in Oversteer.

cazzoo commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,2073 @@
+  {
+    struct t500rs_r41_cmd *r41 = (struct t500rs_r41_cmd *)init_buf;
+    r41->id = 0x41;
+    r41->effect_id = 15; /* Autocenter effect ID */
```
That was actually a good catch. I initially had it enabled at the very beginning (previous driver) and thought it was really necessary to get the forces working but I figured it is still working very fine without it, so removing from the init sequence

cazzoo commented on Makefile:
```diff
@@ -1,6 +1,30 @@
+#   make TMFF2_VERSION=0.1-1
+#
+# Only derive GIT_HASH from git if none was provided and this is a git checkout.
+ifeq ($(origin GIT_HASH), undefined)
```
I found it really useful while working on implementing the driver, making sure the loaded driver is the latest generated. I understand this may be over-complex and doesn't make full sense in several cases, I can definitely remove it if you don't see any benefits

cazzoo replied:
Now I have a pretty decent codebase, I reverted because this is indeed not particularly necessary.

cazzoo commented on docs/T500RS_USB_Protocol_Analysis.md:
```diff
@@ -0,0 +1,663 @@
+
+## Subtype System and Effect Indexing
+
+On this wheel the last six bytes of the 0x01 main upload (bytes 9–14) do **not** contain envelope timings/levels directly. Instead they carry two 16‑bit "subtype" values that act as per‑effect indices:
```
Correct, I'd add that notion

cazzoo commented on docs/T500RS_USB_Protocol_Analysis.md:
```diff
@@ -0,0 +1,663 @@
+
+---
+
+## Important implementation facts
```
Fair, just removed and kept locally for my personal docs

cazzoo committed:
Add square wave (FF_SQUARE) periodic effect support

Add support for FF_SQUARE periodic waveform based on FFEdit USB captures

cazzoo commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,2073 @@
+ * Supported effects.
+ *
+ * NOTE: FF_SQUARE is intentionally OMITTED. Per Windows USB captures, the
+ * T500RS protocol does not encode waveform type in USB packets. Windows/SDL2
```
I managed to get SQUARE wave implemented, with some updated code from ffbsdl and fedit.exe. Will update the PR

cazzoo committed:
Fix 0x05 conditional effect packet structure and scaling + ramp

Corrected packet structure and scaling formulas
Added ramp support

cazzoo commented:
Hi @Kimplul ,

I've been working again on the PR, responding to all your points (almost all -- if not all -- were relevant), and introducing some missing features (FF_SQUARE missing and 0x05 conditional effect that was incorrectly done). If you mind having another pass, I'd be really happy. 
Thanks

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,2073 @@
+
+      t500rs_build_r02_envelope(env, (u8)(env_sub & 0xff), envelope, allow_envelope);
+    }
+      ret =
```
Looks like you missed a few, please check `t500rs_send_packet_sequence` starting with `T500RS_SEQ_ENVELOPE`.

Kimplul commented on src/tmt500rs/hid-tmt500rs.h:
```diff
@@ -0,0 +1,224 @@
+#define T500RS_SAT_INERTIA 100
+
+/* Hardware limits */
+#define T500RS_MAX_EFFECTS 16
```
Now that you're using the ID mapping of `logical_id + 1`, the wheel can only handle 15 effects simultaneously, I think?

cazzoo replied:
Well, The absolute number of effects the wheel supports is 16, constant included. I believe constant can be considered as an effect, and since the logic is doing all the magic to prevent using slot 0 I'd assume we can keep this way.

I you believe we prefer having "MAX_DYNAMIC_EFFECTS" alike variable (set to 15), I'm fine too, it may simplify the logic here or there, but it's literally something we should agree together. Whatever is fine to me, just let me know what suits you the most

Kimplul replied:
The main thing is just that `input_ff_create` in https://github.com/Kimplul/hid-tmff2/blob/26995429a3b92047f0c2f9ab90fe4202c86fd72b/src/hid-tmff2.c#L627 needs to know how many effects can be uploaded to the wheel at a time. I don't think I've seen 16 effects active at the same time, but if such a situation occurs, we want the FFB subsystem to report to the application that it's not supported, rather than send possibly-garbage packets to the wheel.

cazzoo committed:
Added support for boot mode switch to normal mode

cazzoo commented on src/tmt500rs/hid-tmt500rs.h:
```diff
@@ -0,0 +1,224 @@
+  u8 reserved;       /* 0x00 */
+} __packed;
+
+#endif /* __T500RS_PROTOCOL_H */
```
Got it fixed, typo while renaming

cazzoo commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,2073 @@
+ * 90°, 18000 = 180°, etc.)
+ *
+ * Conversion: device_dir = (linux_dir * 36000) / 65536
+ * This maps 0-65535 → 0-35999 (approximately, since 65535 → 35999.45)
```
Good catch, fixed

cazzoo committed:
Minor adjustments (code tidy-up, documentation)

cazzoo commented:
No worries, thank you for the feedback you provided. Take the necessary time for whatever need review. I am not in the hurry, and this driver does work for me already so It can last as long as necessary from your end.

I'm happy we trend to get somewhat a stable mergeable version.
I just pushed a new part of code that handled previously the mode switch (from boot/init to normal mode). This is the part where I have the least confidence. The logic is fully fine, it's working well, but I'm not sure at all on code quality and the compliance/integration with the base driver. If you may give a glance when you can, I'd be happy to send some more time if needed.

Cheers

Kimplul commented:
> EDIT: I just open a PR against hid-tminit for TSCP branch (even though it's t500rs code update)

Yeah, I intended to merge `tspc` into `usb` as part of adding support for the TS-PC, but seems I forgot. I'll sort it out once we deal with this PR, don't want to start changing too many moving bits at the same time.

cazzoo committed:
Reverting init mode switch from driver, it should be handled by hid-tminit

cazzoo committed:
Updated effects slots to be more accurate

cazzoo committed:
Remove un-necessary clamp for direction

Kimplul committed:
rename ffb docs to roughly match

Kimplul committed:
stylistic changes

+ Removal of trailing whitespace, unnecessary unicode chars, small
  formatting changes (though a larger kernel style check should probably
  be performed)

Kimplul commented on docs/T500RS_USB_Protocol_Analysis.md:
```diff
@@ -0,0 +1,574 @@
+
+**Examples:**
+- `01 00 00 40 f4 01 00 00 0e 00 1c 00 00 00` - Constant effect with envelope
+  - Effect ID: 0x00
```
Examples should probably not use ID 0, since the driver doesn't use it.

cazzoo replied:
Corrected

Kimplul commented on docs/T500RS_USB_Protocol_Analysis.md:
```diff
@@ -0,0 +1,574 @@
+
+| Test Case | SDL2 Parameters | Packet Sequence | Packet Payloads |
+|-----------|----------------|-----------------|-----------------|
+| **Triangle Medium** | mag=24000, period=100ms, phase=180°, dir=180° | 1. Upload<br>2. Envelope<br>3. Periodic<br>4. START | `01 00 00 40 d0 07 00 00 0e 00 1c 00 00 00`<br>`02 1c 00 00 12 00 00 12 00`<br>`04 2a 09 00 7f 64 00 00`<br>`41 00 41 01` |
```
While I'm at it, it would maybe be nice if each example not only showed what bytes are sent, but also how they are encoded, like in the docs for the T300. Like

```
0x10 - Effect ID 
0x20 - Effect Code
```

and so on, the current examples require a lot of going back and forth between the different packet formats to figure out what each byte actually means.

Kimplul commented on docs/T500RS_USB_Protocol_Analysis.md:
```diff
@@ -0,0 +1,574 @@
+
+### 10. CONDITIONAL EFFECTS - INERTIA
+
+**Capture Example:**
```
This is a better way to handle examples of each effect upload type, why did the style change midway through the document?

cazzoo replied:
That's historical, but you're right, so I tided and unified the doc format/structure

Kimplul commented on docs/T500RS_USB_Protocol_Analysis.md:
```diff
@@ -0,0 +1,574 @@
+  - SDL 24000 → Device 18
+  - SDL 32767 → Device 255
+
+---
```
Stuff after this line seems to be irrelevant, please remove.

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,1622 @@
+  if (!buf)
+    return -ENOMEM;
+
+  /* Enable autocenter: Report 0x40 0x04 0x01 */
```
Nitpick: structures could be used here as well to give each field a name at least.

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,1622 @@
+    return t500rs_send_hid(t500rs, buf, sizeof(*p));
+  }
+  default:
+    return 0;
```
Shouldn't this be an error since the effect type is either not supported or corrupted?

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,1622 @@
+}
+
+/* Update effect - send parameter updates without re-uploading */
+static int t500rs_update_effect(void *data,
```
Effect updating doesn't seem to handle effect durations/delays or other things that would be part of the `0x01` packet type. Are those unsupported by the wheel or left out to simplify stuff? I think we had a discussion about this but I can't seem to find it at the moment. Either way, that should be documented somewhere. I don't think I've seen modifying effect durations or delays in the wild so it's probably fine, but something to keep in mind.

cazzoo replied:
With T500RS I observed that packets are uploaded once with duration, and then the subsequent updates provides parameter packets that contain forces parameters. If we would like to see duration/delays updated, it would mean a new 0x01 packet to be sent, with a cancellation packet before, and I haven't observed it in my captures. Maybe I missed something but I haven't with all my attempts. Probably real captures from windows would catch this behavior but I don't have and cannot collect by my own anymore.
So, in order to answer you point, I'll add the rationale of this in the code and documentation, just so it becomes explicit.

Kimplul reviewed (changes requested):
Sorry about the delay, I've been busier than expected. I mainly focused on the docs and found some inaccuracies and points of improvement, please have a look at those.

I applied some style fixes myself, please cherry-pick them from https://github.com/Kimplul/hid-tmff2/tree/cazzoo-hardening/t500rs-driver. I don't have push access to your branch and opening a PR for a PR seemed a bit silly, but the changes were trivial enough that I didn't really want to request them explicitly. Kind of a clumsy part of the PR flow, but eh.

Regarding style, I've tried to follow the Linux Kernel style guide in this repo. Most significantly, it dictates a tab width of 8 spaces, whereas this has a width of 2, please fix that. You might also want to try running your code through `clang-format` with the kernel's style configuration, see https://www.kernel.org/doc/html/next/dev-tools/clang-format.html

cazzoo committed:
Code and documentation fixes

- Unicode char replacement
- Documented duration/delays with packet upload
- Code tidy-up (parameters renaming)
- Set structs for init sequence
- Invalid conditional effect types now cause an error instead of silently defaulting
- Ported check for modified parameters to all effects
- Fix saturation handling in T500RS driver & update documentation to reflect it
- Update documentation to clarify dynamic packet code values
- Update examples to use hardware ID 1 instead of 0
- Moved Subtype system section to the bottom
- Hardened documentation style
- Provide details for "DC bias" terminology
- Do not discard 100% gain set but kept warning, more explicit.

cazzoo commented on docs/T500RS_USB_Protocol_Analysis.md:
```diff
@@ -0,0 +1,574 @@
+- Magnitude (0-32767) → Device magnitude (0-127)
+- Phase (0-35999, 0.01° units) → Device phase (0-255, 256 steps for 360°)
+- Period in milliseconds (no conversion needed)
+- Code 0x2a is used (NOT 0x0e as in current driver!)
```
True, there were some inconsistencies, doc updated

cazzoo commented on docs/T500RS_USB_Protocol_Analysis.md:
```diff
@@ -0,0 +1,574 @@
+
+| Test Case | SDL2 Parameters | Packet Sequence | Packet Payloads |
+|-----------|----------------|-----------------|-----------------|
+| **Constant Zero** | level=0, dir=0°, len=500ms | 1. Upload<br>2. Envelope<br>3. Constant<br>4. START | `01 00 00 40 f4 01 00 00 0e 00 1c 00 00 00`<br>`02 1c 00 00 00 00 00 00 00`<br>`03 0e 00 00`<br>`41 00 41 01` |
```
I re-wrote the template as per one of the examples that were fine later in the doc, it should be better now I believe

cazzoo committed:
clang-format pass

Kimplul commented on docs/T500RS_FFBEFFECTS.md:
````diff
@@ -0,0 +1,564 @@
+41 00 41 01 - START command
+```
+- **Packet Type:** 0x41 (Command)
+- **Effect ID:** 0x00 (always 0x00 for T500RS)
````
I believe we've already established that the effect ID varies depending on what effects are loaded, no?

Kimplul commented on docs/T500RS_FFBEFFECTS.md:
````diff
@@ -0,0 +1,564 @@
+41 00 00 01 - STOP command
+```
+- **Packet Type:** 0x41 (Command)
+- **Effect ID:** 0x00 (always 0x00 for T500RS)
````
Ditto.

Kimplul commented on docs/T500RS_FFBEFFECTS.md:
```diff
@@ -0,0 +1,564 @@
+ 4      | 2    | duration_ms    | Duration in milliseconds, little-endian
+ 6      | 2    | delay_ms       | Delay before start, little-endian
+ 8      | 1    | reserved1      | 0x00
+ 9      | 2    | packet_code_1  | Code for subsequent packet type (variable!)
```
`packet_code_1` and `packet_code_2` are the same as `param_sub` and `envelope_sub`, are they not? I think `param_sub` and `envelope_sub` are more descriptive terms, the code uses them as well.

Kimplul commented on docs/T500RS_FFBEFFECTS.md:
```diff
@@ -0,0 +1,564 @@
+
+**IMPORTANT:** Bytes 9-12 specify the packet codes used in subsequent packets. These are NOT fixed values!
+
+**Common Code Combinations:**
```
Again, without explaining subtype indexing this information is not useful to someone reading throught this document for the first time. It's extra noise that makes it more difficult to understand the whole picture.

Kimplul commented on docs/T500RS_FFBEFFECTS.md:
```diff
@@ -0,0 +1,564 @@
+  - Duration: 0x01f4 = 500ms
+  - Delay: 0x0000 = 0ms
+  - Reserved1: 0x00
+  - Packet codes: 0x000e (constant), 0x001c (envelope)
```
I guess it would make more sense to describe how each `param_sub`, `envelope_sub` is calculated (`some_constant + some_other_constant * effect_id`)

Kimplul commented on docs/T500RS_FFBEFFECTS.md:
```diff
@@ -0,0 +1,564 @@
+
+**Code Values:** The code in byte 1 is the low byte of the parameter_subtype from the 0x01 packet (dynamically calculated based on effect index)
+
+**Period Encoding:** Period is in MILLISECONDS (not Hz*100). No conversion needed.
```
Here again, `(not Hz*100)` doesn't seem like important information for a first time reader. Just say the period is in milliseconds, this just makes me want to ask "why is Hz*100 brought up?" which detracts from getting a full picture of what's going on.

Kimplul commented on docs/T500RS_FFBEFFECTS.md:
```diff
@@ -0,0 +1,564 @@
+
+### 1. CONSTANT FORCE EFFECTS
+
+**Capture Examples:**
```
These examples don't seem to even try to explain what's going on. You don't necessarily have to go through each packet field by field, but at least tell the reader what each packet is/does. "Set constant force of zero for 100ms for effect slot 2: Send packet 0x01 with duration 100ms: <packet> Set constant force strength: <packet>` or something along those lines.

Kimplul commented on docs/T500RS_FFBEFFECTS.md:
```diff
@@ -0,0 +1,564 @@
+- Magnitude: 0-127 (scaled from Linux 0-32767)
+- Offset: s8 (-128 to +127, DC bias (Direct Current bias - a constant force offset), scaled from Linux -32768 to +32767)
+- Phase: 0-255 (256 steps for 360 degrees, scaled from Linux 0-35999)
+- Period: Direct milliseconds (no Hz conversion!)
```
What is this `Hz conversion`?

Kimplul commented on docs/T500RS_FFBEFFECTS.md:
```diff
@@ -0,0 +1,564 @@
+- High negative force: `01 00 00 40 88 13 00 00 0e 00 1c 00 00 00` `02 1c 00 00 0d 00 00 0d 00` `03 0e 00 f9`
+- Maximum force with direction: `01 00 00 40 d0 07 00 00 0e 00 1c 00 00 00` `02 1c 00 00 00 00 00 00 00` `03 0e 00 00`
+
+**Packet Structure:**
```
Also, present the format first, then examples.

Kimplul commented on docs/T500RS_FFBEFFECTS.md:
```diff
@@ -0,0 +1,564 @@
+| 6               | 0x00b6            | 0x00c4           |
+
+### Driver Implementation Notes
+- **Effect ID Handling:** The driver uses hardware IDs 1-15 to avoid quirky behavior with hardware index 0 (only valid for constant effects).
```
This is already mentioned above. Sometimes repetition can be useful, but there's only like 15 lines between each occurence so one or the other should probably be removed.

Kimplul commented on docs/T500RS_FFBEFFECTS.md:
```diff
@@ -0,0 +1,564 @@
+- **Device Format:** 16-bit little-endian (0-35999 in 0.01 degree units)
+- **Conversion:** `device_dir = (os_ffb_dir * 36000) / 65536`
+- **Examples:**
+  - 0degrees = 0x0000
```
I believe it's customary to put a space between the number and `degrees`, presumably because `degree` is a whole word. Short forms, such as `ms` for millisecond, can be written without a space between.

Kimplul reviewed (commented):
At least the documentation still needs a bit of work. I would suggest maybe to read through everything a couple of times from a perspective of someone who doesn't know anything about the wheel, reading from top to bottom to get a good overall picture of what the wheel does and how. Present things in the order that they're needed, lots of examples, try to maintain a consistent way to present information, stuff like that.

SeitzB commented:
I have been following this development with some excitement and just tried to build and run the current PR. I do not know how to properly mention files in a comment, but in the hid-tmt500rs-usb.c file there is a capital i after a semicolon.

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,1765 @@
+		/* Wheel handles positive magnitudes only */
+		projected = -projected;
+
+		/* Add 180 degrees to phase to maintain correct force direction.
```
Same here

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,1765 @@
+	offset = (s8)((end_level - start_level) / 512);
+
+	/*
+   * Phase encodes ramp direction per FFEdit captures:
```
Same

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,1765 @@
+	p->id = 0x02;
+	p->subtype = subtype;
+
+	/*
```
Same

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,1765 @@
+		p->fade_len = cpu_to_le16(env->fade_length);
+		p->fade_level = t500rs_scale_envelope_level(env->fade_level);
+	} else {
+		/*
```
Same

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,1765 @@
+			} else if (effect->type == FF_CONSTANT ||
+				   effect->type == FF_PERIODIC) {
+				envelope = &effect->u.periodic.envelope;
+				allow_envelope =
```
This just looks funny :p

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,1765 @@
+		case T500RS_SEQ_CONDITION_X: {
+			const struct ff_condition_effect *cond =
+				&effect->u.condition[0];
+			/* Scale saturation from Linux FFB range (0..65535) to device range
```
Comment spacing issue continues here

Kimplul commented:
@SeitzB I think you have the wrong branch checked out, `master` seems to have the spurious `I` while the most recent changes are in `hardening/t500rs-driver`.

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,1765 @@
+		return -ENOMEM;
+
+	/* Enable autocenter: Report 0x40 0x04 0x01 */
+	{
```
This extra block is unnecessary

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,1765 @@
+
+	T500RS_DBG(t500rs, "Sending initialization sequence...\n");
+
+	/* Report 0x42 - Init/status commands (2 bytes each)
```
More comment whitespace issue

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,1765 @@
+		hid_warn(t500rs->hdev,
+			 "Init command 3 (0x40 config) failed: %d\n", ret);
+
+	/* Report 0x43 - Set global gain (2 bytes)
```
Same

Kimplul commented on src/tmt500rs/hid-tmt500rs.c:
```diff
@@ -0,0 +1,1765 @@
+			break;
+		}
+
+		case T500RS_SEQ_CONDITION_Y: {
```
Does anyone use the second axis? The T300 only uses `effect->u.condition[0]`, and I don't remember seeing kernel drivers use `condition[1]`.

Kimplul reviewed (commented):
I added some comments about the most recent stylistic changes. Looks like `clang-format` did something weird with the whitespacing in multi-line comments, or maybe they were weird to begin with and `clang-format` just choked. Dunno.

cazzoo committed:
docs(T500RS): Address review comments - improve flow, terminology, and examples

Major changes:
- Add QUICK START section for immediate practical learning
- Add Understanding Effects section with real-world analogies
- Reorganize documentation: practice-first, reference-later
- Standardize terminology: param_sub/env_sub instead of packet_code
- Fix degrees spacing and improve parameter descriptions
- Add consistent example format with packet-by-packet breakdowns
- Fix code comment formatting to kernel style
- Remove redundant & 0xff casts when casting to u8
  The cast to u8 already truncates to 8 bits, making & 0xff redundant
- Fix SDL references to Linux FFB subsystem in header file comments
- Verify no unicode characters in source files
- Remove unnecessary braces
- Added helper functions for build&send packets

Kimplul commented:
@hoover67 This should work:

```
git clone --recurse-submodules --branch=hardening/t500rs-driver https://github.com/cazzoo/hid-tmff2.git
```

From there on, you should be able to just follow the installation instructions in the README.

hoover67 commented:
Thanks a lot for these instructions! 

Where should I report errors during compilation? I tried both methods (make and dkms) and both fail with the same compiler error message: 

```
 CC [M]  /home/hoover/src/hid-tmff2/src/tmtspc/hid-tmtspc.o
  CC [M]  /home/hoover/src/hid-tmff2/src/tmt500rs/hid-tmt500rs.o

/home/hoover/src/hid-tmff2/src/tmt500rs/hid-tmt500rs.c: In function ‘t500rs_send_condition_packet’:
/home/hoover/src/hid-tmff2/src/tmt500rs/hid-tmt500rs.c:481:16: error: implicit declaration of function ‘t500rs_send_hid’; did you mean ‘t300rs_send_int’? [-Werror=implicit-function-declaration]
  481 |         return t500rs_send_hid(t500rs, buf,
      |                ^~~~~~~~~~~~~~~
      |                t300rs_send_int

```

and several others following this error which look like version mismatches in function names: 

```
error: implicit declaration of function ‘t500rs_build_r03_constant’; did you mean ‘t500rs_build_r05_condition’? [-Werror=implicit-function-declaration]
  516 |         t500rs_build_r03_constant(r3, code, scaled_level);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~

```

Please forgive my ignorance as this is probably the wrong channel to report these issues in... 

All the best, 

Uwe

hoover67 commented:
Good morning @Kimplul & thanks for the suggestions. 

I got a bit further today by installing Assetto Corsa Rally (a rather new sim) running on "Proton experimental" and I'm happy to report that FFB works nicely and more or less OOTB with this title, so it's definitely some problem with rfactor2. 

I also have eurotruck simulator 2, would it be helpful to see if this title works on my end? 

I tried using these launch options in steam with rfactor2: 

PROTON_ENABLE_HIDRAW=1 SDL_JOYSTICK_HIDAPI=0 

and also went for older Proton versions like 9.0, sadly all to no avail. The wheel & pedals work fine in rfactor2, but not ffb. 

I also tried disabling "SteamInput" which did not make a difference, either. 

During one test session I noticed some "gain set successfully" from the ffb kernel driver (quite high at 55700 or thereabouts), but no force registered on the wheel. So it looks that rf2 is trying to talk to the wheel somehow, just not correctly maybe?

```
Feb 21 10:25:38 dirac kernel: tmff2 0003:044F:B65E.000E: FFB: set_gain 55703 -> device 216
Feb 21 10:25:38 dirac kernel: tmff2 0003:044F:B65E.000E: FFB: Gain set successfully
Feb 21 10:25:51 dirac kernel: tmff2 0003:044F:B65E.000E: FFB: set_gain 55703 -> device 216
Feb 21 10:25:51 dirac kernel: tmff2 0003:044F:B65E.000E: FFB: Gain set successfully
Feb 21 10:27:13 dirac kernel: tmff2 0003:044F:B65E.000E: FFB: set_gain 55703 -> device 216
Feb 21 10:27:13 dirac kernel: tmff2 0003:044F:B65E.000E: FFB: Gain set successfully
Feb 21 10:32:08 dirac kernel: tmff2 0003:044F:B65E.000E: FFB: set_gain 55703 -> device 216


```
Another weird thing is that "oversteer" claims I don't have permissions to change the wheel setup when running it as a normal user; it works fine (obviously) using "sudo", but I'm not sure how the application is supposed to work. The permissions on the /dev/input/event* devices look ok and my user is also in the "input" group.

 Should the wheel radius change immediately when setting it in oversteer? Also, adding a centering force in oversteer makes no difference to how the wheel feels when no simulation is running and just "oversteer" is active. 

I cannot save any profile, however the "tests" section works fine and takes the t500rs through its motions. I am running 0.8.2, maybe I should try cloning the latest repo state (though the last change seems to have been in 2024). 

Again sorry for the noise in this thread, I'll try to register a proper "issue" once it's possible to compile the latest driver state from the regular repo again. 

If there's anything I can provide to help with debugging this feel free to let me know! 

All the best, 

Uwe

cazzoo committed:
Rebase & Fix breaking build

cazzoo commented:
After numerous tests, I haven't found a solution for that issue that happens for both rfactor2 and RRRE. I believe that may be related to Wine/Proton and could be managed by a fix introduced here (but don't know what).
The python script you shared is just resetting the wheel, letting it re-enumerating, and doing mode switch (from boot mode to normal mode). This is handled by the hid-tminit driver, loaded before the hid-tmff driver (which loads then T500RS part). With the latest code, I believe that if you just unplug/replug the wheel, the behavior should be the same as the python script ran.

@Kimplul have you ever heard about that issue. When entering in-game (3D) first, FFB works, when getting to menu FFB drops, and then when getting back to in-game it doesn't work anymore.
The only solution to get FFB back is to re-enumerate the wheel for the game to start effects.

I noticed already during implementation that the games usually don't send start commands. So the wheel is playing once after enumeration, then the packets are being updated in-place (which work for most of the games, but RRRE and Rfactor2).
If the game would eventually stop the effect, there would be then no play effect triggered back, leading to loss of forces.

All of this is assumptions, but this is what I roughly feel

cazzoo commented:
I nailed the issue a bit down and found some more clues. So confirming the issue happens for both games, I found that the game (when doing alt+tab or get back to pits) is stopping effects. When entering back in race, somehow, the effects are not started back..
Here are some debug logs I captures while this issue happens:
```
## pressed alt+tab
[32861.719208] tmff2 0003:044F:B65E.0020: FFB: Effect 0 STOP requested (type=81)
[32861.719318] tmff2 0003:044F:B65E.0020: FFB: Effect 1 STOP requested (type=81)
[32861.719417] tmff2 0003:044F:B65E.0020: FFB: Effect 2 STOP requested (type=85)
[32861.719507] tmff2 0003:044F:B65E.0020: FFB: Effect 3 STOP requested (type=83)
[32861.719592] tmff2 0003:044F:B65E.0020: FFB: Effect 4 STOP requested (type=82)
[32861.719980] tmff2 0003:044F:B65E.0020: FFB: Effect 0 STOP requested (type=81)
[32861.720071] tmff2 0003:044F:B65E.0020: FFB: Effect 0 STOP requested (type=81)
[32861.720075] tmff2 0003:044F:B65E.0020: FFB: Effect 1 STOP requested (type=81)
[32861.720078] tmff2 0003:044F:B65E.0020: FFB: Effect 2 STOP requested (type=85)
[32861.720081] tmff2 0003:044F:B65E.0020: FFB: Effect 3 STOP requested (type=83)
[32861.720084] tmff2 0003:044F:B65E.0020: FFB: Effect 4 STOP requested (type=82)
[32861.720088] tmff2 0003:044F:B65E.0020: FFB: Effect 0 STOP requested (type=81)
[32861.720091] tmff2 0003:044F:B65E.0020: FFB: Effect 1 STOP requested (type=81)
[32861.720094] tmff2 0003:044F:B65E.0020: FFB: Effect 2 STOP requested (type=85)
[32861.720097] tmff2 0003:044F:B65E.0020: FFB: Effect 3 STOP requested (type=83)
[32861.720099] tmff2 0003:044F:B65E.0020: FFB: Effect 4 STOP requested (type=82)
[32861.720310] tmff2 0003:044F:B65E.0020: FFB: Effect 0 STOPPED (type=81)
[32861.723410] tmff2 0003:044F:B65E.0020: FFB: Effect 1 STOPPED (type=81)
## got back ingame (alt+tab), no FFB
[32861.723698] tmff2 0003:044F:B65E.0020: FFB: Effect 0 START requested (count=2147483647, type=81)
[32861.723930] tmff2 0003:044F:B65E.0020: FFB: Effect 1 START requested (count=2147483647, type=81)
[32861.724150] tmff2 0003:044F:B65E.0020: FFB: Effect 2 START requested (count=2147483647, type=85)
[32861.724427] tmff2 0003:044F:B65E.0020: FFB: Effect 3 START requested (count=2147483647, type=83)
[32861.739411] tmff2 0003:044F:B65E.0020: FFB: Effect 2 STARTED (type=85, count=2147483647)
[32861.755421] tmff2 0003:044F:B65E.0020: FFB: Effect 3 STARTED (type=83, count=2147483647)
[32861.771467] tmff2 0003:044F:B65E.0020: FFB: Effect 4 STOPPED (type=82)
```

So I tested adding an extra re-init script (manually triggered) : Stop work handler, Reset all effect states and then Reinitialize wheel (tmff->open), and that worked!

So I don't exactly know if there's a bug to fix in Wine/Proton, but that workaround worked. I'm trying to get it added to the base driver but I'm not entirely convinced this is the right place, especially knowing this is shared code with other wheel bases...

Kimplul commented:
>  So I tested adding an extra re-init script (manually triggered) : Stop work handler, Reset all effect states and then Reinitialize wheel (tmff->open), and that worked!

> So I don't exactly know if there's a bug to fix in Wine/Proton, but that workaround worked. I'm trying to get it added to the base driver but I'm not entirely convinced this is the right place, especially knowing this is shared code with other wheel bases...

I don't think that's a good approach. From the logs it looks like effects 0-3 are all requesting a start, which the driver should be able to handle just fine. Not sure why the game seems to be sending multiple stop signals, but that shouldn't matter.

Looks like all the messages are printed within ~50 milliseconds. Is it maybe possible that the driver handles the STOP requests, and then ignores the START requests because they arrived 'too soon'?

Kimplul commented:
> When a game sends START requests while the work handler is processing effects, a race condition can occur where the START flag is cleared in the first loop but the effect is not actually started until the second loop, causing the effect to never start.
The solution is to batch process the events in a 3 steps sequence.

That doesn't make sense to me. Setting and clearing the flags is done within a spinlock, where exactly would the race condition be?

If you mean that the worker sees a STOP, and while handling the STOP, a START arrives, I could see that the START is not picked up. However, since the worker is running, `delayed_work_pending()` should return false, and the work handler should run a second time, triggering the START. Maybe check if the work handler stops running after an alt-tab? That would be my preferred solution, you solution seems to imply that both REQUEST_START and REQUEST_STOP would be set, which should never be the case.

How is the interrupt/hid messaging stuff related? Is the interupt just slower, meaning a greater chance of mismatched start/stop?

cazzoo commented:
@hoover67 you have to do git fetch cazzoo origin (commands are not correct), then checkout branch "hardening/fix-multi-stop-start-race-condition" from cazzoo origin as well, and then build&install driver.
Recommended, every time you checkout a branch, to do git pull

hoover67 commented:
@cazzoo  Looking great! Just tried rfactor2 with the latest version of your driver including the race condition workaround and I was able to enter the track several times without the wheel losing FFB effects... thanks so much guys for all your hard work.

hoover67 commented:
Hi guys, 

just a quick note to let you know that FFB also works with the RallySimFans Richard Burns Rally mod. 

Also, after updating to Linux Mint 22.3 (was running .0 before), I no longer have to run the tmdrv script above after plugging in the wheel in order for oversteer to detect it correctly (I still have to launch oversteer as root though). 

All the best, 

Uwe

hoover67 commented:
Hi folks, 

I just tried the june-hardening branch and lost all ffb effects on my wheel (in rfactor2, haven't tried other sims yet) except for a rather strong centering force. 

I installed the new version as per instructions in the readme after a full "clone" of the repo. 

i then reverted back to the old driver (thankfully I had a backup) and everything worked as before. 

All the best, 

Uwe

thefahre commented:
Hi all, 

I just tested the driver with /hardening/fix-multi-stop-start-race-condition branch (as @hoover67 said the /hardening/June-pass to not work for him) and force feedback seems to be working, however the force feedback itself feels a bit off, I don't know how to descibe it other than it has very exaggerated force near center yet almost nothing but centering spring when turning. Tried changing settings via oversteer but also doesn't seem to make much of a difference. The game I use to test is Assetto Corsa and Assetto Corsa Competizione.

I'm also unable to validate pedal functionality as I am using Thrustmaster T-LCM pedal directly via USB.

I will try the /hardening/June-pass branch next just to feel the difference.

cazzoo commented:
Hello both,

Thank you for your interest, I am currently working to update the driver and make it work more alike the T300RS and other variants, having effects that lasts length with driver terminating effect.
I will let you know.
Also, if you have double boot, and could capture real session, I'd be really pleased to get them. Thank you

hoover67 commented:
Hey cazzoo, thanks for the info. I still have a win10 partition on my gaming rig somewhere, I have no idea if it still boots up though. 

What exactly would you be interested in when you're asking us to capture a session?

Cheers, Uwe

cazzoo commented:
@hoover67 , I am looking at any game session captures, so I can verify protocol is working as implemented.

Regarding new changes, if you are interested into the latest changes, you can checkout branch [session/agent_a361d90b-d010-4399-8b16-7988ad63f2f8](https://github.com/cazzoo/hid-tmff2/tree/session/agent_a361d90b-d010-4399-8b16-7988ad63f2f8) pulling at it, building and replacing the system driver. Happy to get your feedback on it

thefahre commented:
hi @cazzoo, would a Wireshark capture in bare metal Windows via USBPcap be sufficient? I'm planning to capture a practice session in Assetto Corsa, preparing for a club league endurance event.

thefahre commented:
Hi @cazzoo, following is capture from my T500 RS playing Assetto Corsa in Windows: [Google Drive link](https://drive.google.com/file/d/1YmYorQjLbkDT_YRjZ971x_g70Z-pNaf1/view?usp=sharing)

Also attached are settings of the wheel and the game.
<img width="932" height="899" alt="Screenshot 2026-07-18 120853" src="https://github.com/user-attachments/assets/0a3e8f53-749a-4dae-b7e6-91e2a1adb6f4" />
<img width="932" height="1200" alt="Screenshot 2026-07-18 120819" src="https://github.com/user-attachments/assets/f6110aab-0063-438c-8916-36342e72eb6e" />
<img width="559" height="667" alt="Screenshot 2026-07-18 120755" src="https://github.com/user-attachments/assets/40998f5e-ce07-41bd-b060-b52542b547eb" />
<img width="559" height="667" alt="Screenshot 2026-07-18 120745" src="https://github.com/user-attachments/assets/be8674dd-c51e-441a-a581-cac123941fb1" />

hoover67 commented:
I checked this morning and my win10 install is no longer operational, so I won't be able to provide a USB capture of an rf2 session on short notice... sorry. 

All the best, Uwe

hoover67 commented:
Hi all, 

I managed to revive my win10 partition from the dead and did a little rfactor2 session at Kyalami 1976 in the Brabham BT24 by ChiefWiggum on our practise server (all content is freely available from the steam workshop). 

The pcap file can be found here: 

https://www.icloud.com/iclouddrive/075xh2eBnyMsPO4P34eEm_lmw#t500rs%5Ff1%5Fwheel%5Frfactor2%5Ff1%5F1967%5Fbt24%5Fkyalami%5F1976%5Fonline

I did not do any specific filtering as the t500 wheel is the only device connected to this physical port. (front-facing USB3 port). I had to disable steam input for rfactor2 as the title would not detect any input from my wheel. I left all values (ffb, multipliers etc) at default. 

The session includes plugging in the wheel right at the beginning of the capture (initial rotation of the hardware wheel over the entire range by the firmware). 

Please let me know if any other info is required from my end @cazzoo 

All the best & thanks again for your work on this project! 

Uwe

KORraNpl commented:
hey guys, in case help is needed, I'll do my best. So far I've been dual-booting to play racing games and just tried to use [session/agent_a361d90b-d010-4399-8b16-7988ad63f2f8](https://github.com/cazzoo/hid-tmff2/tree/session/agent_a361d90b-d010-4399-8b16-7988ad63f2f8) branch (I see it's being actively worked on :+1: ), but while the wheel is detected, it jumps by a lot near the center, so it's undrivable in ACC. I've downloaded the most recent Overdrive app and there my T500 RS is not detected at all.

I'm using Kubuntu 24.04 (7.0.0-30-generic), installed module manually via `make && sudo make install`.

kakra commented:
> but while the wheel is detected, it jumps by a lot near the center

Did you try the latest git version and removed old udev rules that may be pre-existing? The current git version has much higher wheel precision. Maybe said branch should be rebased?

thefahre commented:
Gave my feedback to latest changes in #218, commenting here to serve as a link.

hoover67 commented:
Quick update: Tested the new driver in RSF RBR today, appears to work nicely there, too.

cazzoo commented:
Closing this PR, the other one (https://github.com/Kimplul/hid-tmff2/pull/218) does have fresher codebase and better support.
